File: decl-ts/scripts/api-corpus.ts
// The API corpus (tests/api/cases.json) through the high-level API
// (src/api.ts): every case run from the repository root, the answers as
// one JSON array in the form tests/api/README.md fixes — what the parity
// harness diffs across the three drivers (this file, decl-rs/examples/
// api_corpus.rs, decl-py/scripts/api_corpus.py) and what tests/api_test.ts
// compares with tests/api/expected.json.
//
//     node decl-ts/scripts/api-corpus.ts
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { evaluate, check, validate, formatSource, DeclError } from '../src/index.ts';
import type { InputDocument } from '../src/api.ts';

export const root = join(dirname(fileURLToPath(import.meta.url)), '..', '..');

export type Case = {
  name: string;
  evaluate?: string;
  check?: string[];
  validate?: string;
  format_source?: string;
  inputs?: Record<string, { file?: string; json?: unknown }>;
  outputs?: string[];
};
export type Answer =
  | { name: string; ok: true; value: unknown }
  | { name: string; ok: false; message: string; diagnostics: unknown[] };

// a document is a file, named by its path, or the value itself
const documents = (inputs: Case['inputs']): Record<string, InputDocument> =>
  Object.fromEntries(
    Object.entries(inputs ?? {}).map(([k, v]) => [k, 'file' in v ? v.file! : (v.json as any)]),
  );

export async function runCase(c: Case): Promise<Answer> {
  try {
    let value: unknown;
    if (c.evaluate !== undefined) {
      const opts: { inputs?: Record<string, InputDocument>; outputs?: string[] } = {};
      if (c.inputs) opts.inputs = documents(c.inputs);
      if (c.outputs) opts.outputs = c.outputs;
      value = await evaluate(c.evaluate, opts);
    } else if (c.check) value = await check(...c.check);
    else if (c.validate !== undefined)
      value = await validate(c.validate, c.inputs ? { inputs: documents(c.inputs) } : {});
    else if (c.format_source !== undefined) value = await formatSource(c.format_source);
    else throw new Error(`unknown call in ${c.name}`);
    return { name: c.name, ok: true, value };
  } catch (e) {
    if (!(e instanceof DeclError)) throw e;
    return { name: c.name, ok: false, message: e.message, diagnostics: e.diagnostics };
  }
}

/** every case's answer, the cases' paths read from the repository root */
export async function runAll(): Promise<{ cases: Case[]; answers: Answer[] }> {
  process.chdir(root);
  const cases: Case[] = JSON.parse(readFileSync('tests/api/cases.json', 'utf8'));
  const answers: Answer[] = [];
  for (const c of cases) answers.push(await runCase(c));
  return { cases, answers };
}

if (process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1]) {
  console.log(JSON.stringify((await runAll()).answers, null, 2));
}

File: decl-py/scripts/api_corpus.py
"""The API corpus (tests/api/cases.json) through the Python API: every case
run from the repository root, the answers printed as one JSON array in the
form tests/api/README.md fixes — what the parity harness diffs and what
tests/test_api.py compares with tests/api/expected.json.

    decl-py/.venv/bin/python decl-py/scripts/api_corpus.py
"""

from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "decl-py/src"))

import decl


def _document(spec: dict[str, Any]) -> Any:
    # a file is named by its path (a string); any other value is the document itself
    return spec["file"] if "file" in spec else spec["json"]


def run_case(case: dict[str, Any]) -> dict[str, Any]:
    name = case["name"]
    try:
        if "evaluate" in case:
            kw: dict[str, Any] = {}
            if "inputs" in case:
                kw["inputs"] = {k: _document(v) for k, v in case["inputs"].items()}
            if "outputs" in case:
                kw["outputs"] = case["outputs"]
            value: Any = decl.evaluate(case["evaluate"], **kw)
        elif "render" in case:
            kw = {}
            if "inputs" in case:
                kw["inputs"] = {k: _document(v) for k, v in case["inputs"].items()}
            for key in ("outputs", "format", "indent"):
                if key in case:
                    kw[key] = case[key]
            if "templates" in case:
                kw["templates"] = {
                    k: v["file"] if "file" in v else {"text": v["text"]}
                    for k, v in case["templates"].items()
                }
            value = decl.render(case["render"], **kw)
        elif "check" in case:
            value = decl.check(*case["check"])
        elif "validate" in case:
            kw = {}
            if "inputs" in case:
                kw["inputs"] = {k: _document(v) for k, v in case["inputs"].items()}
            value = decl.validate(case["validate"], **kw)
        elif "format_source" in case:
            value = decl.format_source(case["format_source"])
        else:
            raise ValueError(f"unknown call in {name}")
        return {"name": name, "ok": True, "value": value}
    except decl.DeclError as e:
        return {"name": name, "ok": False, "message": str(e), "diagnostics": e.diagnostics}


def run_all() -> list[dict[str, Any]]:
    os.chdir(ROOT)
    cases = json.loads((ROOT / "tests/api/cases.json").read_text(encoding="utf-8"))
    return [run_case(c) for c in cases]


if __name__ == "__main__":
    print(json.dumps(run_all(), indent=2, ensure_ascii=False))
